[Reducer] Move require_finalize_

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100783
* __->__ #100782
* #100781
* #100780

This doesn't need to be set in the loop, just once.

Differential Revision: [D45632426](https://our.internmc.facebook.com/intern/diff/D45632426/)

rohan-varma committed:
[Reducer] Move require_finalize_

This doesn't need to be set in the loop, just once.

Differential Revision: [D45632426](https://our.internmc.facebook.com/intern/diff/D45632426/)

[ghstack-poisoned]

awgu commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -550,13 +550,13 @@ void Reducer::delay_all_reduce() {
 
+  require_finalize_ = true;
   // copy all gradients to buckets
   for (const auto variable_index : c10::irange(params_.size())) {
```
Just to confirm, `params_` is always non-empty?

rohan-varma replied:
It should be, there's somewhere in the reducer that raises an error if DDP is created with no params IIUC. Regardless, finalize_backward() sets this to false upon entry of the function so it should be the same.

rohan-varma committed:
Update on "[Reducer] Move require_finalize_"


This doesn't need to be set in the loop, just once.

Differential Revision: [D45632426](https://our.internmc.facebook.com/intern/diff/D45632426/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI done"
